agda/BinaryRandomAccessList.agda

```diff
@@ -37,21 +40,68 @@ module BinaryRandomAccessList where
   open import Structure.Bin
   open import Container.LeafBinaryTree
 
-  RAL : Set → Set
-  RAL A = DBin Tree
+  RAL-g : Set → ℕ → Set
+  RAL-g A n = DBin-g Tree 
                Container.LeafBinaryTree.toVec Container.LeafBinaryTree.fromVec
-               Container.LeafBinaryTree.iso-to-from Container.LeafBinaryTree.iso-from-to A
+               Container.LeafBinaryTree.iso-to-from Container.LeafBinaryTree.iso-from-to A n
 
+  RAL : Set → Set
+  RAL A = RAL-g A 0
 
   -- TODO: implement all the operations on the datastructure:
   --   cf. "Numerical representation as higher-order nested datatypes", Fig. 3, p.13 for the interface
   --   cf. Okasaki, Sec. 9.2.1 and 10.1.2 for the implementation
 
+  -- Returns an empty structure
   empty : ∀ {A} → RAL A
-  empty = {!!}
-
+  empty = ϵ
+
+  private
+    -- Appends a new Tree in a structure of trees
+    consTree : ∀ {A n} → Tree A n → RAL-g A n → RAL-g A n
+    consTree l ϵ = ϵ [ l ]I
+    consTree l (t O) = t [ l ]I
+    consTree l (t [ l₁ ]I) = consTree (Node l l₁) t O
+
+    -- Removes (and returns) the first tree of the list of trees if not empty
+    unconsTree : ∀ {A n} → RAL-g A n → Maybe $ Tree A n × RAL-g A n
+    unconsTree ϵ = nothing
+    unconsTree (e O) with unconsTree e
+    ... | just (Node t t₁ , ts) = just (t , ts [ t₁ ]I)
+    ... | nothing = nothing
+    unconsTree (e [ t ]I) = just (t , e O)
+
+    -- Retrieves an element in a list of trees
+    lookup-trees : ∀ {A n} → RAL-g A n → ℕ → Maybe A
```

Comment:
Nasty. Would it be easier if we were indexing by a `Bin` instead of `ℕ`?

Reply:
I took the implementations from the document without putting too much though in whether they could be optimized or not. There's a section after that which tries to improve the runtime of the functions which might introduce changes like that but I still have not looked into that. Now that my understanding of the structure is better, I could make such attempts.